libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -224,34 +224,45 @@ string YulUtilFunctions::shiftLeftFunction(size_t _numBits)
 	solAssert(_numBits < 256, "");
 
 	string functionName = "shift_left_" + to_string(_numBits);
-	if (m_evmVersion.hasBitwiseShifting())
-	{
-		return m_functionCollector->createFunction(functionName, [&]() {
-			return
-				Whiskers(R"(
-				function <functionName>(value) -> newValue {
-					newValue := shl(<numBits>, value)
-				}
-				)")
-				("functionName", functionName)
-				("numBits", to_string(_numBits))
-				.render();
-		});
-	}
-	else
-	{
-		return m_functionCollector->createFunction(functionName, [&]() {
-			return
-				Whiskers(R"(
-				function <functionName>(value) -> newValue {
-					newValue := mul(value, <multiplier>)
-				}
-				)")
-				("functionName", functionName)
-				("multiplier", toCompactHexWithPrefix(u256(1) << _numBits))
-				.render();
-		});
-	}
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(value) -> newValue {
+				newValue :=
+				<?hasShifts>
+					shl(<numBits>, value)
+				<!hasShifts>
+					mul(value, <multiplier>)
+				</hasShifts>
+			}
+			)")
+			("functionName", functionName)
+			("numBits", to_string(_numBits))
+			("hasShifts", m_evmVersion.hasBitwiseShifting())
+			("multiplier", toCompactHexWithPrefix(u256(1) << _numBits))
+			.render();
+	});
+}
+
+string YulUtilFunctions::dynamicShiftLeftFunction()
+{
+	string functionName = "shift_left";
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(value, bits) -> newValue {
```

Comment:
Didn't notice that before, but I don't think it's a good idea to reverse the argument order...
Since it's ``shl(bits, value)`` it should also be ``<functionName>(bits, value)``

Reply:
(personally I like ``(value, bits)`` much better and find it more intuitive and I'm always doing it wrong the first time I use shifts myself, but, unfortunately, we can't really change the opcode and we have to stay uniform, otherwise the confusion will be endless)

Reply:
Ah, I didn't notice the differences at all. Fixing